pamu2fcfg/pamu2fcfg.c

```diff
@@ -26,6 +26,198 @@
 #include <readpassphrase.h>
 #endif
 
+static fido_cred_t *prepare_cred(const struct gengetopt_args_info *const args) {
+  fido_cred_t *cred = NULL;
+  fido_opt_t resident_key;
+  char *origin = NULL;
+  char *appid = NULL;
+  char *user = NULL;
+  struct passwd *passwd;
+  unsigned char userid[32];
+  unsigned char cdh[32];
+  char buf[BUFSIZE];
+  int cose_type;
+  int ok = -1;
+  int r;
+
+  if ((cred = fido_cred_new()) == NULL) {
+    fprintf(stderr, "fido_cred_new failed\n");
+    goto err;
+  }
+
+  cose_type = COSE_ES256; /* default */
+  if (args->type_given) {
+    if (!strcasecmp(args->type_arg, "es256")) {
+      cose_type = COSE_ES256;
+    } else if (!strcasecmp(args->type_arg, "rs256")) {
+      cose_type = COSE_RS256;
+    } else {
+      fprintf(stderr, "Unknown COSE type '%s'.\n", args->type_arg);
+      goto err;
+    }
+  }
+
+  if ((r = fido_cred_set_type(cred, cose_type)) != FIDO_OK) {
+    fprintf(stderr, "error: fido_cred_set_type (%d): %s\n", r, fido_strerr(r));
+    goto err;
+  }
+
+  if (!random_bytes(cdh, sizeof(cdh))) {
+    fprintf(stderr, "random_bytes failed\n");
+    goto err;
+  }
+
+  if ((r = fido_cred_set_clientdata_hash(cred, cdh, sizeof(cdh))) != FIDO_OK) {
+    fprintf(stderr, "error: fido_cred_set_clientdata_hash (%d): %s\n", r,
+            fido_strerr(r));
+    goto err;
+  }
+
+  if (args->origin_given) {
+    origin = args->origin_arg;
+  } else {
+    if (!strcpy(buf, PAM_PREFIX)) {
+      fprintf(stderr, "strcpy failed\n");
+      goto err;
+    }
+    if (gethostname(buf + strlen(PAM_PREFIX), BUFSIZE - strlen(PAM_PREFIX)) ==
+        -1) {
+      perror("gethostname");
+      goto err;
+    }
+    origin = buf;
+  }
```

Comment:
Since `buf` is only used here, should we make `origin` an array and ditch `buf`?

Reply:
Makes sense to me. Pushed an pair of commits to do this using `strlcpy()`. Note that the failing tests is the formatting hook which is failing for the imported file (strlcpy.c).